### ISSUE TITLE
L003  over indents - Confused with Functions at CASE STMTs

### ISSUE THREAD

**Issue by OTooleMichael** [sqlfluff/sqlfluff#2853]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Closing / Dedents if all on one line are not understood. As the algo only sees the closing line (the one being computed for faults) as far as the first code char.

```
    SELECT
        MIN(MAX(COALESCE(
            a
        ))) AS c,
        MAX(CASE
            WHEN b THEN 1
            WHEN c THEN 1
        END) as agg
    FROM my_tbl
   ```
   ---> [wrongly corrects to]
   ```
    SELECT
        MIN(MAX(COALESCE(
            a
            ))) AS c,
        MAX(CASE
            WHEN b THEN 1
            WHEN c THEN 1
            END) as agg
    FROM my_tbl
   ```

### Expected Behaviour

In the samle above things are already correct. no need for L003 to do anything.

Under the hood I think the algo is not quite right.
Im guessing it grew organically - After a lot of read and refactoring just to be able to read it. It looks to be half and half using but also subverting the crawler.

### Observed Behaviour

At first I wondered if the over indenting was intentional, and am surprised not to find ti reported, but reading the code it claims to address this exact thing but actually its not possible based on how the code is trying

### How to reproduce

```
    SELECT
        MIN(MAX(COALESCE(
            a
        ))) AS c,
        MAX(CASE
            WHEN b THEN 1
            WHEN c THEN 1
        END) as agg
    FROM my_tbl
``` 

but also any query with many closing brackets on one line.

### Dialect

any

### Version

0.11.0

### Configuration

na

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
I guess this is a matter of opinion but I actually prefer the corrected code. In particular in the `MAX` clause, the `END` this part of the  `CASE` statement and looks weird when it’s unindented.

P.S. I didn’t have anything to do with writing L003 so not biased by that.

**Comment by alanmcruickshank** [comment]
I think this is resolved with #3942.

On that branch the query fixes to:

```sql
SELECT
    MIN(MAX(COALESCE(
        a
    ))) AS c,
    MAX(CASE
        WHEN b THEN 1
        WHEN c THEN 1
    END) as agg
FROM my_tbl
```
